### PR TITLE
add application credentials and OIDC token for OpenStack

### PR DESCRIPTION
**Pull request by zreigz**
**What this PR does / why we need it**: add application credentials and OIDC token for OpenStack

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

```release-note
add application credentials and OIDC token for OpenStack
```

### CODE REVIEW

File: pkg/handler/v1/provider/openstack.go [hunk 1]
```diff
@@ -21,37 +21,51 @@ import (
 	"crypto/x509"
 	"fmt"
 	"net/http"
+	"strings"
+
+	"k8c.io/kubermatic/v2/pkg/handler/middleware"
```
Comment (lsviben) [inline pkg/handler/v1/provider/openstack.go]: NIT: organize imports

File: pkg/provider/cloud/openstack/provider.go [hunk 2]
```diff
@@ -643,73 +646,116 @@ func getNetClientForCluster(cluster kubermaticv1.CloudSpec, dc *kubermaticv1.Dat
...
+	netClient, err := getNetClient(dc.AuthURL, dc.Region, creds, caBundle)
 	if err != nil {
 		return nil, fmt.Errorf("failed to create a authenticated openstack client: %v", err)
 	}
 	return netClient, nil
 }
 
 // GetCredentialsForCluster returns the credentials for the passed in cloud spec or an error
-func GetCredentialsForCluster(cloud kubermaticv1.CloudSpec, secretKeySelector provider.SecretKeySelectorValueFunc) (resources.OpenstackCredentials, error) {
+func GetCredentialsForCluster(cloud kubermaticv1.CloudSpec, secretKeySelector provider.SecretKeySelectorValueFunc) (*resources.OpenstackCredentials, error) {
```
Comment (lsviben) [inline pkg/provider/cloud/openstack/provider.go]: This got really complicated, could you expand the comment here to describe a bit what is the decision flow?

File: pkg/resources/credentials.go [hunk 3]
```diff
@@ -282,6 +285,20 @@ func GetOpenstackCredentials(data CredentialsData) (OpenstackCredentials, error)
...
+	} else if openstackCredentials.ApplicationCredentialID, _ = data.GetGlobalSecretKeySelectorValue(spec.CredentialsReference, OpenstackApplicationCredentialID); openstackCredentials.ApplicationCredentialID != "" {
+		openstackCredentials.ApplicationCredentialSecret, err = data.GetGlobalSecretKeySelectorValue(spec.CredentialsReference, OpenstackApplicationCredentialSecret)
+		if err != nil {
+			return OpenstackCredentials{}, err
+		}
+	}
+
+	if openstackCredentials.ApplicationCredentialID != "" {
+		return openstackCredentials, err
+	}
```
Comment (lsviben) [inline pkg/resources/credentials.go]:
```suggestion
	if spec.ApplicationCredentialID != "" {
		openstackCredentials.ApplicationCredentialID = spec.ApplicationCredentialID
		openstackCredentials.ApplicationCredentialSecret = spec.ApplicationCredentialSecret
		return openstackCredentials, err
	} else if openstackCredentials.ApplicationCredentialID, _ = data.GetGlobalSecretKeySelectorValue(spec.CredentialsReference, OpenstackApplicationCredentialID); openstackCredentials.ApplicationCredentialID != "" {
		openstackCredentials.ApplicationCredentialSecret, err = data.GetGlobalSecretKeySelectorValue(spec.CredentialsReference, OpenstackApplicationCredentialSecret)
		if err != nil {
			return OpenstackCredentials{}, err
		}
		return openstackCredentials, err
	}
```

This should get the same result and is a bit simpler

File: pkg/handler/v1/provider/openstack.go [hunk 4]
```diff
@@ -21,37 +21,51 @@ import (
...
 	userInfo, err := userInfoGetter(ctx, "")
 	if err != nil {
 		return nil, nil, common.KubernetesErrorToHTTPError(err)
 	}
+
+	t := ctx.Value(middleware.RawTokenContextKey)
+	token, ok := t.(string)
+	if !ok || token == "" {
+		return nil, nil, k8cerrors.NewNotAuthorized()
+	}
```
Comment (lsviben) [inline pkg/handler/v1/provider/openstack.go:47]: The token is required, but we only use it if `OIDCAuthentication` is true. What if OIDCAuthentication is not used, do we need to return NotAuthorized if the token is missing?

Comment (lsviben) [reply pkg/handler/v1/provider/openstack.go:47]: Resolved, the token is always present so no issue here.

**Review by mfranczy — APPROVED** [review]
/lgtm

**Pull request merged**